x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/search/InferenceRescorerBuilder.java

```diff
@@ -0,0 +1,274 @@
+        modelWeight = in.readFloat();
+        scoreMode = QueryRescoreMode.readFromStream(in);
+    }
+
+    void setQueryWeight(float queryWeight) {
+        this.queryWeight = queryWeight;
+    }
+
+    void setModelWeight(float modelWeight) {
+        this.modelWeight = modelWeight;
+    }
+
+    void setScoreMode(QueryRescoreMode scoreMode) {
+        this.scoreMode = scoreMode;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (modelSupplier != null) {
+            throw new IllegalStateException("can't serialize model supplier. Missing a rewriteAndFetch?");
+        }
+
+        out.writeString(modelId);
+        out.writeOptionalNamedWriteable(inferenceConfig);
+        boolean fieldMapPresent = fieldMap != null;
+        out.writeBoolean(fieldMapPresent);
+        if (fieldMapPresent) {
+            out.writeMap(fieldMap, StreamOutput::writeString, StreamOutput::writeString);
+        }
+        out.writeFloat(queryWeight);
+        out.writeFloat(modelWeight);
+        scoreMode.writeTo(out);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(MODEL_ID.getPreferredName(), modelId);
+        if (inferenceConfig != null) {
+            builder.startObject(INFERENCE_CONFIG.getPreferredName());
+            builder.field(inferenceConfig.getName(), inferenceConfig);
+            builder.endObject();
+        }
+        if (fieldMap != null) {
+            builder.field(FIELD_MAPPINGS.getPreferredName(), fieldMap);
+        }
+        builder.field(QUERY_WEIGHT.getPreferredName(), queryWeight);
+        builder.field(MODEL_WEIGHT.getPreferredName(), modelWeight);
+        builder.field(SCORE_MODE.getPreferredName(), scoreMode.name().toLowerCase(Locale.ROOT));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public RescorerBuilder<InferenceRescorerBuilder> rewrite(QueryRewriteContext ctx) {
+
+        assert modelId != null;
+
+        if (model != null) {
+            return this;
+        } else if (modelSupplier != null) {
+            if (modelSupplier.get() == null) {
+                return this;
+            } else {
+                return copyScoringSettings(new InferenceRescorerBuilder(modelId, inferenceConfig, fieldMap, modelSupplier.get()));
+
+            }
+        } else {
+            SetOnce<LocalModel> modelHolder = new SetOnce<>();
+
+            ctx.registerAsyncAction(((client, actionListener) -> {
+                TrainedModelProvider modelProvider = new TrainedModelProvider(client, ctx.getXContentRegistry());
+                modelProvider.getTrainedModel(modelId, true, ActionListener.wrap(trainedModel -> {
+                    LocalModel model = new LocalModel(
+                        modelId,
+                        trainedModel.ensureParsedDefinition(ctx.getXContentRegistry()).getModelDefinition(),
```

Comment:
If it is down on the shards, It might be good to not inflate the definition until it is used, that way the query definition is smaller for serializing across the wire. 

I am not sure the computation cost of inflating it once and serializing it vs inflating it X number of times on each shard.

Reply:
👍 The rescorer only runs at the shard. The coordinating node receives the scores from each shard and is responsible for choosing which documents to return but doesn't actually do any rescoring

Reply:
Caching is another consideration. I am not sure if we can cache the inflated object on the shard so that we know not to inflate it and just use the cached object. 

The cost of writing the inflated model on the wire might not be THAT bad if you take into consideration all nodes having to gunzip + parse the JSON every time.